### PR TITLE
[SPARK-13873] [SQL] Avoid copy of UnsafeRow when there is no join in whole stage codegen

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

We need to copy the UnsafeRow since a Join could produce multiple rows from single input rows. We could avoid that if there is no join (or the join will not produce multiple rows) inside WholeStageCodegen.

Updated the benchmark for `collect`, we could see 20-30% speedup.
## How was this patch tested?

existing unit tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 1]
```diff
@@ -396,7 +401,7 @@ case class WholeStageCodegen(child: SparkPlan) extends UnaryNode with CodegenSup
         s"""
            |$evaluateInputs
            |${code.code.trim}
-           |append(${code.value}.copy());
+           |append(${code.value}$doCopy);
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: Is this correct? The row here is created with a `GenerateUnsafeProjection`. So should it be the same row? If so, we should copy the rows even `ctx.copyResult` is false.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 2]
```diff
@@ -378,10 +378,15 @@ case class WholeStageCodegen(child: SparkPlan) extends UnaryNode with CodegenSup
...
+    val doCopy = if (ctx.copyResult) {
+      ".copy()"
+    } else {
+      ""
+    }
     if (row != null) {
       // There is an UnsafeRow already
       s"""
-         |append($row.copy());
+         |append($row$doCopy);
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: Does `ctx.copyResult` only work when the condition `row != null` is true (i.e., child passes a row in)?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: It does not matter.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala (hunk 1 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: If there is only one row will be buffered here, we do not need to copy it. The parent of WholeStageCodegen will do the copy if needed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/GroupedIterator.scala [hunk 3]
```diff
@@ -115,7 +115,8 @@ class GroupedIterator private(
       false
     } else {
       // Skip to next group.
-      while (input.hasNext && keyOrdering.compare(currentGroup, currentRow) == 0) {
+      // currentRow may be overwritten by `hasNext`, so we should compare them first.
+      while (keyOrdering.compare(currentGroup, currentRow) == 0 && input.hasNext) {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/GroupedIterator.scala]: cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 4]
```diff
@@ -78,6 +78,14 @@ class CodegenContext {
   var currentVars: Seq[ExprCode] = null
 
   /**
+   * Whether should we copy the result rows or not.
+   *
+   * If any operator inside WholeStageCodegen generate multiple rows from a single row (for
+   * example, Join), this should be true.
+   */
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I think it's worth nothing that operators which start new pipelines should reset this to false.

**Comment by nongli** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/GroupedIterator.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/GroupedIterator.scala]: LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: understood. LGTM.

**Comment by davies** [comment]
Merged into master.

**Pull request closed without merging**